SimCore/src/SimCore/Reweight/GenieReweightProducer.cxx

```diff
@@ -0,0 +1,203 @@
+#include "RwCalculators/GReWeightAGKY.h"
+#include "RwCalculators/GReWeightDISNuclMod.h"
+#include "RwCalculators/GReWeightDeltaradAngle.h"
+#include "RwCalculators/GReWeightFGM.h"
+#include "RwCalculators/GReWeightFZone.h"
+#include "RwCalculators/GReWeightINuke.h"
+#include "RwCalculators/GReWeightINukeParams.h"
+#include "RwCalculators/GReWeightNonResonanceBkg.h"
+#include "RwCalculators/GReWeightNuXSecCCQE.h"
+#include "RwCalculators/GReWeightNuXSecCCQEaxial.h"
+#include "RwCalculators/GReWeightNuXSecCCQEvec.h"
+#include "RwCalculators/GReWeightNuXSecCCRES.h"
+#include "RwCalculators/GReWeightNuXSecCOH.h"
+#include "RwCalculators/GReWeightNuXSecDIS.h"
+#include "RwCalculators/GReWeightNuXSecNC.h"
+#include "RwCalculators/GReWeightNuXSecNCEL.h"
+#include "RwCalculators/GReWeightNuXSecNCRES.h"
+#include "RwCalculators/GReWeightResonanceDecay.h"
+#include "RwCalculators/GReWeightXSecEmpiricalMEC.h"
+#include "RwCalculators/GReWeightXSecMEC.h"
+#include "RwFramework/GReWeight.h"
+#include "RwFramework/GReWeightI.h"
+#include "RwFramework/GSyst.h"
+#include "RwFramework/GSystSet.h"
+#include "SimCore/Event/EventWeights.h"
+#include "SimCore/Event/HepMC3GenEvent.h"
+
+namespace simcore {
+
+GenieReweightProducer::GenieReweightProducer(const std::string& name,
+                                             framework::Process& process)
+    : Producer(name, process) {
+  hepMC3Converter_ = new genie::HepMC3Converter;
+  genie_rw_ = new genie::rew::GReWeight;
+}
+
+GenieReweightProducer::~GenieReweightProducer() {
+  delete hepMC3Converter_;
+  delete genie_rw_;
+}
+
+void GenieReweightProducer::configure(framework::config::Parameters& ps) {
+  hepmc3CollName_ = ps.getParameter<std::string>("hepmc3CollName");
+  hepmc3PassName_ = ps.getParameter<std::string>("hepmc3PassName");
+  eventWeightsCollName_ = ps.getParameter<std::string>("eventWeightsCollName");
+  verbosity_ = ps.getParameter<int>("verbosity");
+  seed_ = ps.getParameter<int>("seed");
+  n_weights_ = (size_t)(ps.getParameter<int>("n_weights"));
+  auto var_types_strings =
+      ps.getParameter<std::vector<std::string> >("var_types");
+
+  std::default_random_engine generator(seed_);
+  std::normal_distribution<double> normal_distribution;
+
+  for (auto const& vt_str : var_types_strings) {
+    auto vtype = ldmx::EventWeights::string_to_variation_type(vt_str);
+    for (size_t i_w = 0; i_w < n_weights_; ++i_w)
+      variation_map_[vtype].push_back(normal_distribution(generator));
+  }
+}
+
+void GenieReweightProducer::reinitializeGenieReweight() {
+  delete genie_rw_;
+  genie_rw_ = new genie::rew::GReWeight;
+
+  genie::RunOpt::Instance()->SetTuneName(tune_);
+  if (!genie::RunOpt::Instance()->Tune()) {
+    EXCEPTION_RAISE("ConfigurationException", "No TuneId in RunOption.");
+  }
+  genie::RunOpt::Instance()->BuildTune();
+
+  genie_rw_->AdoptWghtCalc("hadro_fzone", new genie::rew::GReWeightFZone);
+  genie_rw_->AdoptWghtCalc("hadro_intranuke", new genie::rew::GReWeightINuke);
```

Comment:
Does `genie_rw_` own these objects you created with `new`? (will it `delete` them?)

Reply:
Yes (or at least it claims to own), and yes, it does delete them.